File: src/core/ops/consumer.ts
import type { ConsumerDef, Registry } from "../../registry/types.ts";
import { MenvError } from "../errors.ts";
import type { OpResult } from "./util.ts";
import { cloneRegistry, newPlan, requireConsumer, requireSlug, requireVault } from "./util.ts";

export interface ConsumerAddInput {
  name: string;
  strategyType: "single" | "per-vault";
  baseDir: string;
  filename?: string;
  filenames?: Record<string, string>;
  secretsAsLocalOverrides?: boolean;
  example?: boolean;
}

function buildDef(registry: Registry, input: ConsumerAddInput): ConsumerDef {
  const flags = {
    ...(input.secretsAsLocalOverrides ? { secretsAsLocalOverrides: true } : {}),
    ...(input.example ? { example: true } : {}),
  };
  if (input.strategyType === "single") {
    if (input.filename === undefined) {
      throw new MenvError("VALIDATION", "single strategy needs --filename");
    }
    return { strategyType: "single", strategyConfig: { baseDir: input.baseDir, filename: input.filename, ...flags } };
  }
  if (input.filenames === undefined || Object.keys(input.filenames).length === 0) {
    throw new MenvError("VALIDATION", "per-vault strategy needs --filenames <vault>=<file>,…");
  }
  for (const vault of Object.keys(input.filenames)) requireVault(registry, vault);
  return {
    strategyType: "per-vault",
    strategyConfig: { baseDir: input.baseDir, filenames: input.filenames, ...flags },
  };
}

export function planConsumerAdd(registry: Registry, input: ConsumerAddInput): OpResult {
  requireSlug("consumer", input.name);
  if (registry.consumers[input.name] !== undefined) {
    throw new MenvError("VALIDATION", `consumer "${input.name}" already exists`);
  }
  const def = buildDef(registry, input);
  const next = cloneRegistry(registry);
  next.consumers[input.name] = def;
  const plan = newPlan();
  plan.registry.push({
    action: "set",
    path: `consumers.${input.name}`,
    summary: `add consumer "${input.name}" (${input.strategyType}, ${input.baseDir})`,
  });
  return { next, plan };
}

export interface ConsumerUpdateInput {
  name: string;
  baseDir?: string;
  filename?: string;
  filenames?: Record<string, string>;
  secretsAsLocalOverrides?: boolean;
  example?: boolean;
}

export function planConsumerUpdate(registry: Registry, input: ConsumerUpdateInput): OpResult {
  const def = requireConsumer(registry, input.name);
  if (def.strategyType === "per-vault" && input.filename !== undefined) {
    throw new MenvError("VALIDATION", `"${input.name}" is per-vault — use --filenames, not --filename`);
  }
  if (def.strategyType === "single" && input.filenames !== undefined) {
    throw new MenvError("VALIDATION", `"${input.name}" is single — use --filename, not --filenames`);
  }
  if (input.filenames !== undefined) {
    for (const vault of Object.keys(input.filenames)) requireVault(registry, vault);
  }
  const next = cloneRegistry(registry);
  const target = next.consumers[input.name] as ConsumerDef;
  const changed: string[] = [];
  if (input.baseDir !== undefined) {
    target.strategyConfig.baseDir = input.baseDir;
    changed.push("baseDir");
  }
  if (target.strategyType === "single" && input.filename !== undefined) {
    target.strategyConfig.filename = input.filename;
    changed.push("filename");
  }
  if (target.strategyType === "per-vault" && input.filenames !== undefined) {
    target.strategyConfig.filenames = { ...target.strategyConfig.filenames, ...input.filenames };
    changed.push("filenames");
  }
  if (input.secretsAsLocalOverrides !== undefined) {
    target.strategyConfig.secretsAsLocalOverrides = input.secretsAsLocalOverrides;
    changed.push("secretsAsLocalOverrides");
  }
  if (input.example !== undefined) {
    target.strategyConfig.example = input.example;
    changed.push("example");
  }
  const plan = newPlan();
  if (changed.length > 0) {
    plan.registry.push({
      action: "set",
      path: `consumers.${input.name}`,
      summary: `update consumer "${input.name}" (${changed.join(", ")})`,
    });
  }
  return { next, plan };
}

export interface ConsumerRemoveInput {
  name: string;
  // Vaults a session could be opened for; orphaned keys elsewhere become a
  // warning (Plan 3's `check` reports lingering keys), never a blocker.
  openable: Set<string>;
  // The consumer's generated paths (main + .local + .env.example), computed by
  // the caller via consumerPaths. Released (disclaimer stripped) by default, or
  // deleted with --delete-files. Marker-guarded at apply time. OPTIONAL so the
  // Plan-2 unit tests (registry cascade only) keep calling without them.
  paths?: string[];
  deleteFiles?: boolean;
}

// Registry cascade + orphan-key cleanup. Generated-file release/deletion is
// Plan 3 (no generate yet, so no files exist to manage).
export function planConsumerRemove(registry: Registry, input: ConsumerRemoveInput): OpResult {
  requireConsumer(registry, input.name);
  const next = cloneRegistry(registry);
  const plan = newPlan();
  delete next.consumers[input.name];
  plan.registry.push({
    action: "remove",
    path: `consumers.${input.name}`,
    summary: `remove consumer "${input.name}"`,
  });

  const lockedVaultsWithOrphans = new Set<string>();
  for (const [varName, def] of Object.entries(next.variables)) {
    for (const [vault, byConsumer] of Object.entries(def.vaultMapping)) {
      const entry = byConsumer[input.name];
      if (entry === undefined) continue;
      delete byConsumer[input.name];
      plan.registry.push({
        action: "remove",
        path: `variables.${varName}.vaultMapping.${vault}.${input.name}`,
        summary: `unwire "${varName}" from "${input.name}" (vault "${vault}")`,
      });
      const stillUsed = Object.values(byConsumer).some((e) => e.key === entry.key);
      if (!stillUsed) {
        if (input.openable.has(vault)) {
          plan.vaults.push({ vault, action: "remove", key: entry.key });
        } else {
          lockedVaultsWithOrphans.add(vault);
        }
      }
      if (Object.keys(byConsumer).length === 0) delete def.vaultMapping[vault];
    }
  }
  for (const vault of [...lockedVaultsWithOrphans].sort()) {
    plan.warnings.push({
      code: "ORPHANED_KEYS",
      message: `vault "${vault}" could not be opened — keys orphaned by removing "${input.name}" remain (menv check will report them)`,
    });
  }
  for (const path of input.paths ?? []) {
    plan.files.push({ action: input.deleteFiles === true ? "delete" : "release", path });
  }
  return { next, plan };
}

File: go/internal/core/ops/consumer.go
package ops

import (
	"fmt"
	"sort"

	"github.com/nikrabaev/menv/go/internal/core"
	"github.com/nikrabaev/menv/go/internal/registry"
)

// ConsumerAddInput holds the parameters for PlanConsumerAdd.
type ConsumerAddInput struct {
	Name                    string
	StrategyType            string // "single" | "per-vault"
	BaseDir                 string
	Filename                string            // single strategy
	Filenames               map[string]string // per-vault strategy
	SecretsAsLocalOverrides bool
	Example                 bool
}

func buildConsumerDef(r registry.Registry, input ConsumerAddInput) (registry.ConsumerDef, error) {
	cfg := registry.StrategyConfig{
		BaseDir: input.BaseDir,
	}
	if input.SecretsAsLocalOverrides {
		cfg.SecretsAsLocalOverrides = true
	}
	if input.Example {
		cfg.Example = true
	}
	switch input.StrategyType {
	case "single":
		if input.Filename == "" {
			return registry.ConsumerDef{}, &core.MenvError{
				Code:    core.ErrValidation,
				Message: "single strategy needs --filename",
			}
		}
		cfg.Filename = input.Filename
	case "per-vault":
		if len(input.Filenames) == 0 {
			return registry.ConsumerDef{}, &core.MenvError{
				Code:    core.ErrValidation,
				Message: "per-vault strategy needs --filenames <vault>=<file>,…",
			}
		}
		for vault := range input.Filenames {
			if _, err := RequireVault(r, vault); err != nil {
				return registry.ConsumerDef{}, err
			}
		}
		cfg.Filenames = input.Filenames
	default:
		return registry.ConsumerDef{}, &core.MenvError{
			Code:    core.ErrValidation,
			Message: fmt.Sprintf("unknown strategy type %q (use single or per-vault)", input.StrategyType),
		}
	}
	return registry.ConsumerDef{StrategyType: input.StrategyType, StrategyConfig: cfg}, nil
}

// PlanConsumerAdd registers a new consumer.
func PlanConsumerAdd(r registry.Registry, input ConsumerAddInput) (OpResult, error) {
	if err := RequireSlug("consumer", input.Name); err != nil {
		return OpResult{}, err
	}
	if _, exists := r.Consumers[input.Name]; exists {
		return OpResult{}, &core.MenvError{
			Code:    core.ErrValidation,
			Message: fmt.Sprintf("consumer %q already exists", input.Name),
		}
	}
	def, err := buildConsumerDef(r, input)
	if err != nil {
		return OpResult{}, err
	}
	next := CloneRegistry(r)
	next.Consumers[input.Name] = def
	plan := NewPlan()
	plan.Registry = append(plan.Registry, core.RegistryOp{
		Action:  "set",
		Path:    fmt.Sprintf("consumers.%s", input.Name),
		Summary: fmt.Sprintf("add consumer %q (%s, %s)", input.Name, input.StrategyType, input.BaseDir),
	})
	return OpResult{Next: next, Plan: plan}, nil
}

// ConsumerUpdateInput holds the parameters for PlanConsumerUpdate.
type ConsumerUpdateInput struct {
	Name                    string
	BaseDir                 *string           // nil = no change
	Filename                *string           // nil = no change (single only)
	Filenames               map[string]string // nil = no change (per-vault only)
	SecretsAsLocalOverrides *bool             // nil = no change
	Example                 *bool             // nil = no change
}

// PlanConsumerUpdate modifies an existing consumer's configuration.
func PlanConsumerUpdate(r registry.Registry, input ConsumerUpdateInput) (OpResult, error) {
	def, err := RequireConsumer(r, input.Name)
	if err != nil {
		return OpResult{}, err
	}
	if def.StrategyType == "per-vault" && input.Filename != nil {
		return OpResult{}, &core.MenvError{
			Code:    core.ErrValidation,
			Message: fmt.Sprintf("%q is per-vault — use --filenames, not --filename", input.Name),
		}
	}
	if def.StrategyType == "single" && len(input.Filenames) > 0 {
		return OpResult{}, &core.MenvError{
			Code:    core.ErrValidation,
			Message: fmt.Sprintf("%q is single — use --filename, not --filenames", input.Name),
		}
	}
	if len(input.Filenames) > 0 {
		for vault := range input.Filenames {
			if _, err := RequireVault(r, vault); err != nil {
				return OpResult{}, err
			}
		}
	}

	next := CloneRegistry(r)
	target := next.Consumers[input.Name]
	var changed []string

	if input.BaseDir != nil {
		target.StrategyConfig.BaseDir = *input.BaseDir
		changed = append(changed, "baseDir")
	}
	if target.StrategyType == "single" && input.Filename != nil {
		target.StrategyConfig.Filename = *input.Filename
		changed = append(changed, "filename")
	}
	if target.StrategyType == "per-vault" && len(input.Filenames) > 0 {
		if target.StrategyConfig.Filenames == nil {
			target.StrategyConfig.Filenames = map[string]string{}
		}
		for k, v := range input.Filenames {
			target.StrategyConfig.Filenames[k] = v
		}
		changed = append(changed, "filenames")
	}
	if input.SecretsAsLocalOverrides != nil {
		target.StrategyConfig.SecretsAsLocalOverrides = *input.SecretsAsLocalOverrides
		changed = append(changed, "secretsAsLocalOverrides")
	}
	if input.Example != nil {
		target.StrategyConfig.Example = *input.Example
		changed = append(changed, "example")
	}
	next.Consumers[input.Name] = target

	plan := NewPlan()
	if len(changed) > 0 {
		sort.Strings(changed)
		plan.Registry = append(plan.Registry, core.RegistryOp{
			Action:  "set",
			Path:    fmt.Sprintf("consumers.%s", input.Name),
			Summary: fmt.Sprintf("update consumer %q (%s)", input.Name, joinComma(changed)),
		})
	}
	return OpResult{Next: next, Plan: plan}, nil
}

// ConsumerRemoveInput holds the parameters for PlanConsumerRemove.
type ConsumerRemoveInput struct {
	Name        string
	Openable    map[string]bool
	Paths       []string // generated file paths (optional; present in CLI layer)
	DeleteFiles bool
}

// PlanConsumerRemove removes a consumer and cascades to all variable wirings.
func PlanConsumerRemove(r registry.Registry, input ConsumerRemoveInput) (OpResult, error) {
	if _, err := RequireConsumer(r, input.Name); err != nil {
		return OpResult{}, err
	}
	next := CloneRegistry(r)
	plan := NewPlan()

	delete(next.Consumers, input.Name)
	plan.Registry = append(plan.Registry, core.RegistryOp{
		Action:  "remove",
		Path:    fmt.Sprintf("consumers.%s", input.Name),
		Summary: fmt.Sprintf("remove consumer %q", input.Name),
	})

	lockedWithOrphans := map[string]bool{}
	varNames := make([]string, 0, len(next.Variables))
	for n := range next.Variables {
		varNames = append(varNames, n)
	}
	sort.Strings(varNames)

	for _, varName := range varNames {
		def := next.Variables[varName]
		vaultList := make([]string, 0, len(def.VaultMapping))
		for v := range def.VaultMapping {
			vaultList = append(vaultList, v)
		}
		sort.Strings(vaultList)
		for _, vault := range vaultList {
			byConsumer := def.VaultMapping[vault]
			entry, exists := byConsumer[input.Name]
			if !exists {
				continue
			}
			delete(byConsumer, input.Name)
			plan.Registry = append(plan.Registry, core.RegistryOp{
				Action:  "remove",
				Path:    fmt.Sprintf("variables.%s.vaultMapping.%s.%s", varName, vault, input.Name),
				Summary: fmt.Sprintf("unwire %q from %q (vault %q)", varName, input.Name, vault),
			})
			// If no other consumer shares this key, it's orphaned.
			stillUsed := false
			for _, e := range byConsumer {
				if e.Key == entry.Key {
					stillUsed = true
					break
				}
			}
			if !stillUsed {
				if input.Openable[vault] {
					plan.Vaults = append(plan.Vaults, core.VaultOp{Vault: vault, Action: "remove", Key: entry.Key})
				} else {
					lockedWithOrphans[vault] = true
				}
			}
			if len(byConsumer) == 0 {
				delete(def.VaultMapping, vault)
			}
		}
		next.Variables[varName] = def
	}

	lockedList := make([]string, 0, len(lockedWithOrphans))
	for v := range lockedWithOrphans {
		lockedList = append(lockedList, v)
	}
	sort.Strings(lockedList)
	for _, vault := range lockedList {
		plan.Warnings = append(plan.Warnings, core.PlanIssue{
			Code:    "ORPHANED_KEYS",
			Message: fmt.Sprintf("vault %q could not be opened — keys orphaned by removing %q remain (menv check will report them)", vault, input.Name),
		})
	}

	for _, path := range input.Paths {
		action := "release"
		if input.DeleteFiles {
			action = "delete"
		}
		plan.Files = append(plan.Files, core.FileOp{Action: action, Path: path})
	}

	return OpResult{Next: next, Plan: plan}, nil
}
